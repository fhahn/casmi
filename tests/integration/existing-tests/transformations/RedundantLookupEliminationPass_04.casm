// transform, cmdline "--ir --cf --rle"

CASM RedundantLookupEliminationPass

init foo

function BLOCK : -> Int initially { 500 }

rule foo =
{|
	BLOCK := undef

    call (@bb_beq)(0x80001f58)// -> 648
    // --------------------------------
    // let l0__ = 10 in
    //   let l1__ = 10 in
    //     let l2__ = 648 in
	//       if l0__ = l1__ then
	// 	       BLOCK := l2__
    
	if BLOCK!=648 and  true then 
    {|
        assert( false )
	|}
    
	if BLOCK=650 then 
    {|
        assert( false )
    |}
    
	if BLOCK=648 then 
    {|
        BLOCK:=654
	    call (@bb_bne)(0x80001f98)// -> 653
        // --------------------------------
        // let l3__ = 20 in
        //   let l4__ = 30 in
        //     let l5__ = 653 in
	    //       if l3__ = l4__ then
		//         BLOCK := l5__
	|}
	
    if BLOCK=654 then 
    {|
        BLOCK:=651
    	call (@bb_j)(0x80001fa0)// -> 54
	    // -----------------------------
        // let l6__ = 54 in
	    //   BLOCK := l6__
    |}

    program(self) := undef
|}

rule bb_beq(addr : Int) =
let rs = 10 in
let rt = 10 in
let offset = 648 in
	if rs = rt then
		BLOCK := offset

rule bb_bne(addr : Int) =
let rs = 20 in
let rt = 30 in
let offset = 653 in
	if rs = rt then
		BLOCK := offset

rule bb_j(addr : Int) =
let _addr = 54 in
	BLOCK := _addr


