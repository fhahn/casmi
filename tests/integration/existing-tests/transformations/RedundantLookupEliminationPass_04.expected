foo=STATEMENTS(LET CONST(ident c0__,num 10),STATEMENTS(LET CONST(ident c1__,num 648),STATEMENTS(LET CONST(ident c2__,FALSE),STATEMENTS(LET CONST(ident c3__,num 650),STATEMENTS(LET CONST(ident c4__,num 654),STATEMENTS(LET CONST(ident c5__,num 20),STATEMENTS(LET CONST(ident c6__,num 30),STATEMENTS(LET CONST(ident c7__,num 653),STATEMENTS(LET CONST(ident c8__,num 651),STATEMENTS(LET CONST(ident c9__,num 54),STATEMENTS(LET CONST(ident c10__,TRUE),SEQBLOCK(STATEMENTS(STATEMENTS(LOCAL(ident v1__,UNDEF),UPDATE(ident BLOCK,ident v1__)),STATEMENTS(STATEMENTS(SKIP,STATEMENTS(PARBLOCK(PARBLOCK(PARBLOCK(IF(ident c10__,UPDATE(ident BLOCK,ident c1__)),null),null),null),null)),STATEMENTS(IF(ident c10__,SEQBLOCK(STATEMENTS(ASSERT(ident c2__,null),null),null)),STATEMENTS(IF(ident c2__,SEQBLOCK(STATEMENTS(ASSERT(ident c2__,null),null),null)),STATEMENTS(IF(ident c2__,SEQBLOCK(STATEMENTS(UPDATE(ident BLOCK,ident c4__),STATEMENTS(STATEMENTS(PARBLOCK(PARBLOCK(PARBLOCK(IF(ident c2__,UPDATE(ident BLOCK,ident c7__)),null),null),null),null),null)),null)),STATEMENTS(IF(=(ident BLOCK,ident c4__),SEQBLOCK(STATEMENTS(UPDATE(ident BLOCK,ident c8__),STATEMENTS(STATEMENTS(PARBLOCK(UPDATE(ident BLOCK,ident c9__),null),null),null)),null)),STATEMENTS(UPDATE(LOOKUP(ident program,LIST(SELF,null)),UNDEF),null))))))),null))))))))))))

bb_beq=STATEMENTS(LET CONST(ident c0__,num 10),STATEMENTS(LET CONST(ident c1__,num 648),STATEMENTS(LET CONST(ident c2__,TRUE),PARBLOCK(PARBLOCK(PARBLOCK(IF(ident c2__,UPDATE(ident BLOCK,ident c1__)),null),null),null))))

bb_bne=STATEMENTS(LET CONST(ident c0__,num 20),STATEMENTS(LET CONST(ident c1__,num 30),STATEMENTS(LET CONST(ident c2__,num 653),STATEMENTS(LET CONST(ident c3__,FALSE),PARBLOCK(PARBLOCK(PARBLOCK(IF(ident c3__,UPDATE(ident BLOCK,ident c2__)),null),null),null)))))

bb_j=STATEMENTS(LET CONST(ident c0__,num 54),PARBLOCK(UPDATE(ident BLOCK,ident c0__),null))
